crates/walrus-sdk/src/config/communication_config.rs

```diff
@@ -33,6 +35,134 @@ fn default_sliver_status_check_threshold() -> usize {
     DEFAULT_SLIVER_STATUS_CHECK_THRESHOLD
 }
 
+/// Default number of sliver uploads that should be observed before evaluating throughput.
+pub const DEFAULT_AUTO_TUNE_WINDOW_SAMPLE_TARGET: usize = 10;
+/// Default timeout while waiting for a throughput sample window to complete.
+pub const DEFAULT_AUTO_TUNE_WINDOW_TIMEOUT: Duration = Duration::from_secs(10);
+/// Default multiplicative factor applied when exploring higher permit counts.
+pub const DEFAULT_AUTO_TUNE_INCREASE_FACTOR: f64 = 1.5;
+/// Default lock factor applied to the best-performing permit count.
+pub const DEFAULT_AUTO_TUNE_LOCK_FACTOR: f64 = 0.9;
+/// Default minimum number of permits that the auto-tuner evaluates.
+pub const DEFAULT_AUTO_TUNE_MIN_PERMITS: usize = 50;
+/// Default maximum number of permits that the auto-tuner evaluates.
+pub const DEFAULT_AUTO_TUNE_MAX_PERMITS: usize = 1_000;
+/// Default weight applied to secondary slivers relative to primaries when computing throughput.
+pub const DEFAULT_AUTO_TUNE_SECONDARY_WEIGHT: f64 = 0.5;
+
+fn default_auto_tune_window_sample_target() -> usize {
+    DEFAULT_AUTO_TUNE_WINDOW_SAMPLE_TARGET
+}
+
+fn default_auto_tune_window_timeout() -> Duration {
+    DEFAULT_AUTO_TUNE_WINDOW_TIMEOUT
+}
+
+fn default_auto_tune_increase_factor() -> f64 {
+    DEFAULT_AUTO_TUNE_INCREASE_FACTOR
+}
+
+fn default_auto_tune_lock_factor() -> f64 {
+    DEFAULT_AUTO_TUNE_LOCK_FACTOR
+}
+
+fn default_auto_tune_min_permits() -> usize {
+    DEFAULT_AUTO_TUNE_MIN_PERMITS
+}
+
+fn default_auto_tune_max_permits() -> usize {
+    DEFAULT_AUTO_TUNE_MAX_PERMITS
+}
+
+fn default_auto_tune_secondary_weight() -> f64 {
+    DEFAULT_AUTO_TUNE_SECONDARY_WEIGHT
+}
+
+/// Configuration for runtime auto-tuning of the data-in-flight limit.
+#[serde_as]
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+#[serde(default)]
+pub struct DataInFlightAutoTuneConfig {
+    /// Enables the additive-increase / multiplicative-decrease controller.
+    pub enabled: bool,
+    /// Number of completed sliver uploads required to close a measurement window.
+    #[serde(default = "default_auto_tune_window_sample_target")]
```

Comment:
Nit: AFAIU, these individual `serde(default)` annotations are not needed if the whole struct is annotated with `serde(default)`.